### DISCUSSION TITLE
Not able to add slider Question in Observations and surveys

### DISCUSSION THREAD

**Discussion by ramkumar-ns-dev** [Sunbird-inQuiry/Community#78] [Q&A]
**Dockstaging**
interactionTypes is one of the property of Question object type
```
"interactionTypes":{
      "type": "array",
      "items": {
        "type": "string",
        "enum": [
          "choice",
          "text",
          "select",
          "date",
          "file-upload",
          "canvas"
        ]
      }
    }
```
We have created following primary categories with target type as Questions :-
    • Multiselect Multiple Choice Question (It has interactionTypes choice)
    • Date (It has interactionTypes date)
    • Slider (It has interactionTypes slider)
    • Text(It has interactionTypes text)
Now when we are creating Multiselect Multiple Choice Question, Date and text it's not throwing error because these primary categories has  interactionTypes which is already available in the Question object.
But in case of slider the interactionTypes it is not available in the Question object hence it's throwing an error but it should not throw error because when we created the slider primary category we have given slider as interactionTypes and you can verify that in the below API.

**Get Object category defenition CURL**
```
curl 'https://dockstaging.sunbirded.org/action/object/category/definition/v1/read?fields=objectMetadata,forms,name,label' \
  -H 'Accept: application/json' \
  -H 'Accept-Language: en-GB,en-US;q=0.9,en;q=0.8' \
  -H 'Authorization;' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Cookie: connect.sid=s%3AumzwJlc3m5ZWeAbRSs7lUfL_vsGinU2C.sHO89zwLnsfsD8DQrnlt%2BYZdLOaXub%2FCbge21OwhGq0' \
  -H 'Origin: https://dockstaging.sunbirded.org' \
  -H 'Referer: https://dockstaging.sunbirded.org/contribute/program/94367220-e348-11ed-87b4-9feca80ba862' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36' \
  -H 'X-App-Id: staging.dock.portal' \
  -H 'X-Device-ID: 7e1cc819b3dd02976a83074a30b21dda' \
  -H 'X-Source: web' \
  -H 'X-msgid: ddcc37d3-d776-e651-d9b9-b62906ac9470' \
  -H 'sec-ch-ua: "Chromium";v="112", "Google Chrome";v="112", "Not:A-Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'ts: 2020-12-31T16:23:27+05:30' \
  --data-raw '{"request":{"objectCategoryDefinition":{"objectType":"Question","name":"Slider"}}}' \
  --compressed
```

**Add Slider Question Request CURL**

```
curl 'https://dockstaging.sunbirded.org/action/questionset/v1/hierarchy/update' \
  -X 'PATCH' \
  -H 'Accept: application/json' \
  -H 'Accept-Language: en-GB,en-US;q=0.9,en;q=0.8' \
  -H 'Authorization;' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Cookie: connect.sid=s%3AumzwJlc3m5ZWeAbRSs7lUfL_vsGinU2C.sHO89zwLnsfsD8DQrnlt%2BYZdLOaXub%2FCbge21OwhGq0' \
  -H 'Origin: https://dockstaging.sunbirded.org' \
  -H 'Referer: https://dockstaging.sunbirded.org/contribute/program/94367220-e348-11ed-87b4-9feca80ba862' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36' \
  -H 'X-App-Id: staging.dock.portal' \
  -H 'X-Device-ID: 7e1cc819b3dd02976a83074a30b21dda' \
  -H 'X-Source: web' \
  -H 'X-msgid: 180a6096-b7b2-23eb-8162-d05b0a087dac' \
  -H 'sec-ch-ua: "Chromium";v="112", "Google Chrome";v="112", "Not:A-Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'ts: 2020-12-31T16:23:27+05:30' \
  --data-raw '{"request":{"data":{"nodesModified":{"93870500-c59c-3d73-8e0b-98e930dbf5b2":{"metadata":{"mimeType":"application/vnd.sunbird.question","media":[],"editorState":{"question":"<p>slider question 1</p>"},"body":"<p>slider question 1</p>","responseDeclaration":{"response1":{"type":"integer","maxScore":1}},"solutions":[],"createdBy":"b717deee-4251-4161-9aae-cd73bf9055f3","author":"rakesh matam","channel":"01329314824202649627","framework":"ekstep_ncert_k-12","license":"CC BY 4.0","name":"slider question title","showEvidence":"No","showRemarks":"No","interactions":{"validation":{"required":"Yes"},"response1":{"validation":{"range":{"min":"100","max":"1000"}},"step":"5"}},"hints":{"en":[null]},"instructions":{"en":[null]},"interactionTypes":["slider"],"primaryCategory":"Slider"},"objectType":"Question","root":false,"isNew":true}},"hierarchy":{"do_21378784932733747211576":{"name":"survey 1","children":["do_21378784951833395211577"],"root":true},"do_21378784951833395211577":{"name":"Section 1","children":["do_21378785012047052811579","do_21378810318271283211585","do_21378872066144665611589","93870500-c59c-3d73-8e0b-98e930dbf5b2"],"root":false}}}}}' \
  --compressed
```

**Error Response For Add slider Question API**

``
{"id":"api.questionset.hierarchy.update","ver":"3.0","ts":"2023-05-04T11:09:49ZZ","params":{"resmsgid":"1506b5a2-2b1a-42c1-bbf3-e180c6cd289d","msgid":null,"err":"CLIENT_ERROR","status":"failed","errmsg":"Validation Errors"},"responseCode":"CLIENT_ERROR","result":{"messages":["Metadata interactionTypes should be one of: [choice, text, select, date, file-upload, canvas]"]}}
``

Can anybody help on this. [@pallakartheekreddy](https://github.com/pallakartheekreddy) [@sajeshkayyath](https://github.com/sajeshkayyath)

CC: [@kiranharidas187](https://github.com/kiranharidas187)

**Comment by krgauraw** [comment]
@ramkumar-pacewisdom Can you please restart `assessment` micro-service? Sometimes local in-memory cache won't get cleared.

cc: @sajeshkayyath / @kiranharidas187

**Reply by VISHNUDAS-tunerlabs** [reply]
@krgauraw We are able to locate this job, can you please confirm if we can run this job to restart the micro-service : http://10.20.0.14:8080/jenkins/job/ArtifactUpload/job/DockStaging/job/Core/job/Assessment/
cc: @ramkumar-pacewisdom @githubpiyush

**Comment by ramkumar-ns-dev** [comment]
@krgauraw thank you, we are checking and will update here.

**Comment by ramkumar-ns-dev** [comment]
@krgauraw we've restarted the below services, But seems to be not working

`http://10.20.0.14:8080/jenkins/job/ArtifactUpload/job/DockStaging/job/Core/job/Assessment/`
`http://10.20.0.14:8080/jenkins/job/Deploy/job/DockStaging/job/Kubernetes/job/Assessment/`

Can you please check this once ? 

CC: @aks30 @sajeshkayyath @kiranharidas187

**Reply by ramkumar-ns-dev** [reply]
@krgauraw can you help us with alternate solution ?

**Reply by ramkumar-ns-dev** [reply]
@sajeshkayyath can i get any help in this anyways?

**Reply by krgauraw** [reply]
@ramkumar-pacewisdom service restart is not guaranteed when deploying job with no change in the build. 
The correct job for assessment service deployment is http://10.20.0.14:8080/jenkins/job/Deploy/job/DockStaging/job/Kubernetes/job/Assessment/

So Can you please get the service restarted with the help of devops team or you can attempt fresh build and deploy 

cc: @aks30 @sajeshkayyath @kiranharidas187 @sajeshkayyath @PoojaJManjooran @kameshbhr

**Reply by ramkumar-ns-dev** [reply]
@krgauraw for the above job link we've tried restart assessment service and also tried with fresh build as well, but it's not working. Would you please suggest any other ways we can try ?

**Reply by krgauraw** [reply]
@ramkumar-pacewisdom this should work if the restart happens successfully. Can you please get it done with devops team (try deleting the pod and force restart) .
I have created Slider category and its definition under inquiry dev environment and it's working fine. You may use this env to test your feature.

Hope this helps!

cc: @sajeshkayyath @kameshbhr @PoojaJManjooran

**Reply by ramkumar-ns-dev** [reply]
@krgauraw Thank you, will check with devops team for the force restart and regarding the slider category will check in dev inquiry.

**Reply by ramkumar-ns-dev** [reply]
@krgauraw i tried by deleting the pod and force restart with help of devops team and it's seems to be not working. would you suggest any other way ?
